### ISSUE TITLE
Add a header propagation middleware and message handler

### ISSUE THREAD

**Issue by rynowak** [aspnet/HttpClientFactory#190]
We should consider providing an in-the-box solution for propagating headers from the current request to an outgoing http request.

This would consist of a middleware that can be configured to *stamp* certain headers on a feature of the current request. Then a message handler uses `IHttpContextAccessor` to grab the feature and apply the headers to outgoing requests. 

Maybe similar to:
https://github.com/stevejgordon/CorrelationId/tree/dev/src/CorrelationId from @stevejgordon or 
(sample @davidfowl made)

Also consider using `Activity`

**Comment by rynowak** [comment]
Marking for up for grabs. If you're interested in this please continue the discussion here first so we can work out details.

**Comment by martincostello** [comment]
I’d be potentially interested in helping out with this. We currently have some in-house bespoke handlers to do header forwarding, so a configurable, built-in way to do the same thing would be great.

**Comment by stevejgordon** [comment]
Hi @rynowak - Depending on timescales I'd quite like to throw in on this one since you may be after similar to my library. I've been meaning to dig into the Activity bits at some point too. I've just recently added a handler within an internal project that applies the correlation ID using my library which works quite nicely.

**Comment by stevejgordon** [comment]
Ah, @martincostello beat me to the punch! Clearly we're both watching notifications from this repo!! :-)

**Comment by martincostello** [comment]
I’m sure multiple heads thinking about it are better than one!

**Comment by rynowak** [comment]
cool - I'd definitely look forwards to some contributions from you both. I think I'd also like to get some input from @davidfowl and @glennc and use this issue to discuss design proposals as well as scenarios.

**Comment by davidfowl** [comment]
Here's the sample I wrote https://gist.github.com/davidfowl/c34633f1ddc519f030a1c0c5abe8e867. This is just generic header propagation not tied to any specific implementation of activities or id generation etc.

**Comment by stevejgordon** [comment]
I was thinking the exact same thing as @martincostello. It's something we recently needed to do and a few people asked about adding that to the CorrelationId library too. Also, some people ask about being able to control this per endpoint. In that respect, something that can be added per logical client on the HttpClientBuilder would be useful too.

**Comment by davidfowl** [comment]
@stevejgordon 

> Also, some people ask about being able to control this per endpoint. In that respect, something that can be added per logical client on the HttpClientBuilder would be useful too.

See https://gist.github.com/davidfowl/c34633f1ddc519f030a1c0c5abe8e867#file-headerpropagation-cs-L24

> Would simple transformations be on the table for this feature as well? We do a few things internally where we map one header to another name when we pass it on.

As for the mapping I'd like to see a couple of real examples. I believe there's a need there but I don't know the extent of the mapping capabilities required. Is it a simple `Func<IncomingHeader, OutgoingHeader>`? Or do you need more context? Maybe there's an "I want to take over" mode where you get to write the headers yourself:


```C#
services.AddHeaderPropagation(options =>
{
        options.Mapping = (httpContext, httpRequestMessage) =>
        {
                // Write code to propagate headers here
        };
});
```

**Comment by martincostello** [comment]
I like this idea of it sort of being "layered" so you can go low-level and custom if you really want to.

Something super-simple could then be powered off that with a helper overload, something kinda like:

```csharp
public static IServiceProvider AddHeaderPropagation(this IServiceProvider services, NameValueCollection mapping)
{
    return services.AddHeaderPropagation(options =>
    {
        options.Mapping = (context, request) =>
        {
            foreach (string key in mapping.Keys)
            {
                var headerValue = context.Request.Headers[key];

                if (StringValues.IsNullOrEmpty(headerValue))
                {
                    continue;
                }

                request.Headers.TryAddWithoutValidation(mapping[key], (string[])headerValue);
            }
        };
    });
}
```

Then in the simple case, a user can do something like the below:

```csharp
var mapping = new NameValueCollection()
{
    ["Referer"] = "Referer",
    ["User-Agent"] = "X-Forwarded-User-Agent",
    ["X-Correlation-Id"] = "X-Correlation-Id",
};

services.AddHeaderPropagation(mapping);
```

But for a more advanced implementation, the user can just bring their own implementation and be advanced as they like using the `Action<HttpContext, HttpRequestMessage>` delegate.

**Comment by martincostello** [comment]
So internally we map headers so that we can trace things in our Kibana logs so we can see where traffic originated as it goes through the layers of our application, such as from our DDoS edge, to HAProxy, to our web tier, down to our microservices/APIs.

In one case we have a header called `x-je-conversation` which we always pass through if present that is sort of a end-to-end correlation ID, which would be covered by the simple propagation case.

In another case we map `User-Agent` to `x-je-user-agent` so that an application can see who called it via `User-Agent` and who it called it on behalf of from `x-je-user-agent`. This would be covered by the "mapping" scenario.

We also have a third case where we pass-through a header if present, but generate a new value if not to forward on. This would be something more advanced that would either still need a custom handler today, or could use something like the advanced delegate-based mapper.

**Comment by glennc** [comment]
I'm not arguing that we don't do this feature. But if the majority of customer cases are satisfied by correlationIds and perhaps some baggage then we should possibly focus our efforts there first.

For example, a scenario that would need mapping outside of the tracing system is testing in production where you flow a header that informs some load balancer to flow your requests to a specific test instance of a microservice.

**Comment by martincostello** [comment]
Yep, it's definitely similar to/is-kinda distributed tracing, it's just something custom and internal that's been around for years and years in our many applications that everything has to follow by convention. 

At the moment for our .NET Core stuff we're doing this semi-manually with a hand-coded delegating handler we distribute via ProGet, but something built-in we could just configure would be nicer, however that manifests itself.

I think the key "want" for that scenario for us is being able to map one header already present to another name, rather than only a simple as-is propagation.
